### ISSUE TITLE
indexes - owner, times, size, extension

### ISSUE THREAD

**Issue by bmcgough** [FredHutch/storage-crawler#8]
second pass: uid, gid, mode(to find files/links/dirs),path/dir

**Comment by dtenenba** [comment]
If this is a note to add indexes to the `owner` column (and other columns) in the `file_metadata` table (or whatever table that view is derived from), then I would like to +1 it. (If not let me know and I'll create a separate issue).

I am working on a little web app to show directories with a lot of data in them (based on example 1
[here](https://docs.google.com/document/d/1IvzWOgbeQDLxkTlWX0bg2lcQBmxvba8mAaZgy0Ymxck/edit#)) and I'm finding that some queries are really slow. For example:

```sql
SELECT file_metadata.owner,  uid_mapping.name, file_metadata.filename, 
to_timestamp(file_metadata.mtime), file_metadata.sum FROM file_metadata
JOIN uid_mapping ON file_metadata.uid = uid_mapping.uid
order by file_metadata.sum desc limit 10;
```

This takes over 10 minutes (I gave up waiting at a certain point). 

If I add a `WHERE owner = 'peters_u'` clause it speeds up to 6.3 seconds. But I'd like to do this query (in a reasonable time) both with and without the `where` clause (my real query is a little more involved but this simplified version illustrates the problem).

Actually I am not sure that problem can be fixed just by adding an index to `owner`. 

But this one probably can: 

```sql
select distinct owner from file_metadata order by owner;
```

That takes about 137 seconds.

So it would be great if an index could be added to the `owner` column. 

I need this query to populate a dropdown in my little app so people can choose different PIs. @dirkpetersen suggested getting the PI list from https://toolbox.fhcrc.org/json/pi_all.json but that list does not contain `_ADM` and other owners that show up with the query above.

He also suggested a newer version of postgres which can do queries in parallel. That plus the index would probably speed things up quite a bit. If you like I can make a PR which creates the indexes...

**Comment by bmcgough** [comment]
Oh, geeze! PostgreSQL - go figure!

The owner column _is_ indexed. Longer story short, if the anticipated scan, using the index, is a certain ratio of the table, then the index is not used. Using an index is two lookups for each value, so only really helps when you are looking up fewer than half the rows.

But the `distinct` issue is different. PostgreSQL's `distinct` function is old and very poorly optimized. This is kind of an interesting exercise...

- `select distinct owner from file_metadata` takes about 150000ms
- `with recursive fm(n) as (select min(owner) from file_metadata union all select(select file_metadata.owner from file_metadata where file_metadata.owner > n order by file_metadata.owner limit 1) from fm where n is not null) select n from fm;` takes about 180ms

WHAT?!?!?

As for your first query, what you want is an index on file_metadata.sum. I've done that by hand for today's results, and the query now returns in 17ms for a limit of 10. I will add that index to the nightly script.

**Comment by bmcgough** [comment]
Current indicies:

owner
atime,mtime
inode,parent_inode
fileextension
st_mode
sum

**Comment by bmcgough** [comment]
Forgot to explain that query.

MySQL and other RDMSes do a 'skip' thing when selecting distinct, which PostgreSQL does not. This recursive query should skip over already found owners by min()-sorting them. This then causes it to use the owner index very efficiently.

**Comment by dtenenba** [comment]
This is what I get when running that query:

```
storcrawldb=> with recursive fm(n) as (select min(owner) from file_metadata union all select(select file_metadata.owner from file_metadata where file_metadata.owner > n order by file_metadata.owner limit 1) from f where n is not null);
ERROR:  syntax error at or near ";"
LINE 1: ...by file_metadata.owner limit 1) from f where n is not null);
                                                                      ^
```

If I remove the semicolon from the end it does not give an error, but also gives no results.
???

**Comment by bmcgough** [comment]
Sorry, left off the last part of the query. Above comment has been fixed and tested.
